File: go/utils/counter_test.go
package utils

import (
	"net/http"
	"testing"
)

func TestFilterHeadersFunc(t *testing.T) {

	confidentialHeaderSlice := []string{"Authorization", "Cookie", "X-Request-ID"}

	headers := http.Header{}
	headers.Add("testHeader", "testHeader")
	for _, header := range confidentialHeaderSlice {
		headers.Add(header, "testHeader")
	}

	filteredHeaders := FilterHeaders(headers)

	if len(filteredHeaders) == 0 {
		t.Error("filteredHeaders should not be empty")
	}

	for i := 0; i < len(confidentialHeaderSlice); i++ {
		if filteredHeaders[confidentialHeaderSlice[i]] != nil {
			t.Error("filteredHeaders should not contain confidential headers")
		}
	}

}

func TestCalculateHeaderByteLength(t *testing.T) {

	headers := http.Header{}

	headers.Add("testHeader", "testHeader")
	headers.Add("testHeader", "testHeader2")

	bytes := CalculateHeaderByteLength(headers)

	if bytes != 31 {
		t.Error("bytes should be 31 -- ", bytes)

	}

}

File: ts/src/utils/counter.test.ts
import { MockResponseWriter } from "../testutils";
import { HttpHeader } from "../types";
import { CalculateHeaderByteLength, CountingReadableStream, CountingResponseWriter, FilterHeaders } from "./counter";
import { stream2string } from "./stream2string";

describe("test filter", () => {
  it("CountingRequestReader", async () => {
    const rs = new CountingReadableStream(
      new ReadableStream<Uint8Array>({
        start(controller) {
          const encoder = new TextEncoder();
          controller.enqueue(encoder.encode("Hello"));
          controller.enqueue(encoder.encode(" "));
          controller.enqueue(encoder.encode("World"));
          controller.enqueue(encoder.encode("!"));
          controller.close();
        },
      }),
    );
    const res = await stream2string(rs);
    expect(res).toBe("Hello World!");
    expect(rs.ReadBytes).toBe(res.length);
  });

  it("CountingResponseWriter", async () => {
    const mw = new MockResponseWriter();
    const cw = new CountingResponseWriter(mw);
    cw.WriteHeader(200);
    cw.Header().Add("testHeader", "testHeader");
    await cw.Write("Hello World!");
    await cw.Write("Hello World!");
    await cw.End();
    expect(mw.Body).toBe("Hello World!Hello World!");
    expect(cw.WrittenBytes).toBe(24);
  });

  it("FilterHeaders", () => {
    const confidentialHeaderSlice = ["Authorization", "Cookie", "X-Request-ID"];
    const headers = new HttpHeader();

    headers.Add("testHeader", "testHeader");
    for (const header of confidentialHeaderSlice) {
      headers.Add(header, "testHeader");
    }
    const filteredHeaders = FilterHeaders(headers);
    expect(filteredHeaders.Items().length).toBe(1);
    for (const ch of confidentialHeaderSlice) {
      const f = filteredHeaders.Get(ch);
      expect(f).toBe(undefined);
    }
  });
  it("CalculateHeaderByteLength", () => {
    const headers = new HttpHeader();
    headers.Add("testHeader", "testHeader");
    headers.Add("testHeader", "testHeader2");
    const bytes = CalculateHeaderByteLength(headers);

    expect(bytes).toBe(49);
  });
});
